service/service.go

```diff
@@ -177,38 +177,38 @@ func (s *Service) loop(ctx context.Context) {
 	var executingRounds errgroup.Group
 	defer executingRounds.Wait()
 
-	timer := time.NewTimer(0)
-	<-timer.C
-	defer timer.Stop()
 	for {
-		start := s.genesis.
-			Add(s.cfg.EpochDuration * time.Duration(s.openRound.Epoch())).
-			Add(s.cfg.PhaseShift)
-		if d := time.Until(start); d > 0 {
-			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), d)
-			timer.Reset(d)
-			select {
-			case <-timer.C:
-			case <-ctx.Done():
-				log.Info("service shutting down")
-				s.openRoundMutex.Lock()
-				s.openRound = nil
-				s.openRoundMutex.Unlock()
-				return
-			}
+		s.openRoundMutex.RLock()
+		epoch := s.openRound.Epoch()
+		s.openRoundMutex.RUnlock()
+
+		start := s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift)
+		waitTime := time.Until(start)
+		if waitTime > 0 {
+			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), waitTime)
+		}
+		select {
+		case <-time.After(waitTime):
+		case <-ctx.Done():
+			log.Info("service shutting down")
+			s.openRoundMutex.Lock()
+			s.openRound = nil
+			s.openRoundMutex.Unlock()
+			return
```

Comment:
There is a leaking go routine here:

```suggestion
		timer := time.NewTimer(time.Until(s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift))
		select {
		case <-timer.C:
		case <-ctx.Done():
			log.Info("service shutting down")
			if !timer.Stop() {
				<-timer.C
			}
			s.openRoundMutex.Lock()
			s.openRound = nil
			s.openRoundMutex.Unlock()
			return
```

Reply:
The proposed code:
```go
timer := time.NewTimer(time.Until(s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift))
```

has exactly the same issue as the original. The calculations are just squeezed into one line. Avoiding a local variable `waitTime` doesn't prevent the problem. 

Where is the leaking goroutine? If you refer to garbage collection of the `Timer`  inside of `time.After()` - it is a non-issue. It is GCed after the timer hits. The only "leak" is when the application closes before the timer ticks, but it's not a problem as the app closes anyway.

Reply:
I am aware that the one liner is no different than calculating `waitTime` and then starting the timer.

The point is not to calculate `waitTime`, then doing other stuff (logging in this case) and then starting the timer.

Reply:
Done